### PR TITLE
Ast2 27  users add background layers

### PR DESCRIPTION
**Pull request by Hornebom**

### CODE REVIEW

File: src/client/components/layer-list/layer-list.vue [hunk 1]
```diff
@@ -69,11 +81,25 @@ export default {
...
     setExpanded(id) {
       this.expanded = this.expanded !== id ? id : ''
     },
+    calculateMaxHeight() {
+      /**
+       * calculating possible max-height
+       * magic numbers: 300 as distance to top, 80 as spacing at bottom
+       */
+      // TODO/Extra: Add resize handler to recalculate maxHeight
+      const top = this.$refs.list ? this.$refs.list.$el.getBoundingClientRect().top : 300;
```
Comment (petergoes) [inline src/client/components/layer-list/layer-list.vue]: Please write a ternary if statement like so:
```
condition
 ? true value
 : false value
```

It makes it more readable then having it on one line

File: src/client/components/layer-list/layer-list.vue [hunk 2]
```diff
@@ -1,16 +1,20 @@
 <template>
-  <md-list>
+  <md-list ref="list" :style="{maxHeight: maxHeight + 'px'}">
     <li
-      v-for="layer in layers"
-      :key="layer.id"
+      v-for="(layer, index) in layers"
+      :key="layer.id + index"
```
Comment (petergoes) [inline src/client/components/layer-list/layer-list.vue]: why is the `layer.id` not enough and is the index needed?

File: src/client/components/layer-dialog/layer-dialog.vue [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  props: {
+    showLayerDialog: {
+      type: Boolean,
+      default: false,
+    },
+  },
+  data() {
+    return {
+      serverUrl: 'test',
+      serverType: 'MOCK',
```
Comment (petergoes) [inline src/client/components/layer-dialog/layer-dialog.vue]: `'test'` and `'MOCK'` look like development values? Can they be removed?

File: src/client/components/layer-dialog/layer-dialog.vue [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      state: 'settings', //TODO: Look if working with this state is the best way to go
```
Comment (petergoes) [inline src/client/components/layer-dialog/layer-dialog.vue]: Is this `TODO` followed up?

File: src/client/store/data/wmsLayers.js [hunk 5]
```diff
@@ -1,8 +1,10 @@
 import getData from '../../lib/get-data'
+import { getRealApiData } from "../../lib/get-api-data";
 export const state = () => []
 
 export const mutations = {
   addLayer(state, layer) {
+    // console.log('layer', layer)
```
Comment (petergoes) [inline src/client/store/data/wmsLayers.js]: Please clean up

File: src/client/store/data/wmsLayers.js [hunk 6]
```diff
@@ -15,6 +17,34 @@ export const actions = {
...
+  },
+
+  addCustomTilesToLayers({ state, commit, dispatch }, customLayers) {
+    const newLayers = customLayers.map(source => {
+      const id = source.id ? source.id : source.name.split(" ")[0]
+      const mapLayer = {
+        id: id,
+        title: source.name,
+        layerType: 'raster',
+        url: source.tiles[0], // TODO: source.tiles should be returned from the backend as url (String)
```
Comment (petergoes) [inline src/client/store/data/wmsLayers.js]: Is this `TODO` being followed up?

File: src/client/store/project.js [hunk 7]
```diff
@@ -100,6 +100,7 @@ export const mutations = {
     state.settings.targets[group][key] = { ...state.settings.targets[group][key], ...value }
   },
   setWmsLayer(state, layer) {
+    // TODO: Only add if layer not already in layers (check id)
```
Comment (petergoes) [inline src/client/store/project.js]: Is this `TODO` being followed up?

File: src/client/components/layer-list/layer-list.vue (hunk 2 above)
Comment (Hornebom) [reply src/client/components/layer-list/layer-list.vue]: Leftover from unfinished api and same id's. removed

File: src/client/components/layer-dialog/layer-dialog.vue [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+      const customLayersIds = this.customLayers.map(layer => layer.id)
+      return this.layers.map(layer => customLayersIds.includes(layer.id) );
+    },
+    layerError() {
+      return this.layers.map(layer => layer.errors.length > 0 );
+    },
+  },
+  mounted() {
+    // this.serverType = this.options[0].name
+  },
```
Comment (petergoes) [inline src/client/components/layer-dialog/layer-dialog.vue]: This method can be removed

File: src/client/components/layer-dialog/layer-dialog.vue [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+  display: flex;
+  margin-top: var(--spacing-default);
+}
+
+.layer-dialog__layer-label {
+  width: 100%;
+}
+
+.layer-dialog__layer-disabled {
+  /* text-decoration: line-through; */
```
Comment (petergoes) [inline src/client/components/layer-dialog/layer-dialog.vue]: Why is this line disabled?

File: src/client/components/layer-dialog/layer-dialog.vue [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+
+.layer-dialog__layer .md-list-item-content {
+  align-items: flex-start;
+  padding-top: 14px;
+}
+
+.layer-dialog__layer-hint {
+  width: 100%;
+  margin-top: 5px;
+  font-size: 14px;
```
Comment (petergoes) [inline src/client/components/layer-dialog/layer-dialog.vue]: Should be: `font-size: var(--font-size-small);`

File: src/client/components/layer-dialog/layer-dialog.vue [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+  display: inline-block;
+  vertical-align: top;
+  width: 18px;
+  height: 18px;
+  min-width: 18px;
+  margin-top: -2px;
+  margin-right: 4px;
+
+  font-size: 16px !important;
+  line-height: 16px;
```
Comment (petergoes) [inline src/client/components/layer-dialog/layer-dialog.vue]: Use variables here

File: src/client/components/layer-list/layer-list.vue [hunk 12]
```diff
@@ -8,17 +8,31 @@
         :class="{ 'md-active': expanded === layer.id }"
         class="md-list-item-expand md-list-item-container"
       >
+        <template v-if="layer.errors">
+          <p v-for="(error, index) in layer.errors" :key="index">layer has errors</p>
```
Comment (petergoes) [inline src/client/components/layer-list/layer-list.vue]: Translate

File: src/client/components/layer-list/layer-list.vue [hunk 13]
```diff
@@ -86,4 +158,69 @@ export default {
...
+
+.layer-list__avatar {
+  background-color: var(--neutral-color--light);
+}
+
+.layer-list__avatar-fake {
+  width: 40px;
+  height: 40px;
+  min-width: 40px;
+  margin-right: 16px;
```
Comment (petergoes) [inline src/client/components/layer-list/layer-list.vue]: Use variables

File: src/client/components/layer-list/layer-list.vue [hunk 14]
```diff
@@ -86,4 +158,69 @@ export default {
...
+
+.layer-list__add-btn {
+  width: 100%;
+  height: auto;
+  margin: 0;
+  padding: 0;
+}
+
+.layer-list__add-btn .md-ripple {
+  padding: 12px 16px 12px 24px;
```
Comment (petergoes) [inline src/client/components/layer-list/layer-list.vue]: Use variables

File: src/client/components/map-box/map-box.vue [hunk 15]
```diff
@@ -113,6 +121,12 @@ export default {
       this.clearMap()
       this.$nextTick(this.fillMap)
     },
+    wmsLayers() {
+      [...this.wmsLayers].reverse().forEach(this.addWmsLayer)
+    },
+    customLayers() {
+      [...this.customLayers].reverse().forEach(this.addWmsLayer)
+    },
```
Comment (petergoes) [inline src/client/components/map-box/map-box.vue]: Should the reverse not be done in `allMapLayers`?

File: src/client/store/data/customLayers.js [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+  },
+}
+
+export const actions = {
+  async getCustomMapLayers({ commit, dispatch }, parameters) {
+    const body = {
+      url: parameters.serverUrl,
+      type: parameters.serverType,
+    }
+    const data = await getRealApiData('maplayers', body)
```
Comment (petergoes) [inline src/client/store/data/customLayers.js]: `getRealApiData` is an implementation detail, should be an other method available

File: src/client/store/data/wmsLayers.js [hunk 17]
```diff
@@ -1,4 +1,6 @@
 import getData from '../../lib/get-data'
+import randomColor from '../../lib/random-color';
+import { getRealApiData } from "../../lib/get-api-data";
```
Comment (petergoes) [inline src/client/store/data/wmsLayers.js]: Where are these imports used?

File: src/client/store/project.js [hunk 18]
```diff
@@ -294,6 +304,15 @@ export const actions = {
       commit('setWmsLayer', { id: layer.id, visible: false, showLegend: false, opacity: 1 })
     })
   },
+  bootstrapCustomLayers({ state, commit }, layers) {
+    layers.forEach(layer => {
+      commit('setCustomLayer', {
+        ...layer,
+        visible: false,
+        showLegend: false,
+        opacity: 1 })
```
Comment (petergoes) [inline src/client/store/project.js]: `})` to next line

File: src/client/store/project.js [hunk 19]
```diff
@@ -587,4 +606,9 @@ export const getters = {
       opacity,
     }))
   },
+
+  customLayers: (state, getters, rootState, rootGetters) => {
+    console.log(rootState)
```
Comment (petergoes) [inline src/client/store/project.js]: remove log statement

**Pull request merged**